### PR TITLE
Graphs use plotly for data visualisation

### PR DESCRIPTION
**Pull request by TomDijkema**
Pull request contains new functionalities for graphical support, also the overlap between GBIF and GeoCASe data has been built into this request.

Adds:
- Request functions for requesting data from stored csv files
- Support for GBIF and GeoCASe data (inside the request functions) to be merged
- Drawing functions for creating bar and line charts using Plotly based on the stored data

Changes:
- Folder structure of csv files is now organized per month and stored in: 'csv_files/storage', instead of: 'csv_files/written'
- Changed the naming conventions of csv files and functions in main.py to be more in line with their respected infrastructures (GBIF and GeoCASe)
- Moved function for creating issues and flags list to: query_csv.py

Removes:
- Draft functions for creating graphs with Plotly

### CODE REVIEW

File: query_csv.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
+import csv
+import copy
+from itertools import islice
+from datetime import datetime as dt
+
+
+current_month = dt.now().strftime('%B')
+
+
+def read_country_data(month=current_month):
```
Comment (samleeflang) [inline query_csv.py]: This is a really long function. Try to split it up and give them descriptive name so it is clear what they are doing.

File: GBIF_functions.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+from typing import Any
```
Comment (samleeflang) [inline GBIF_functions.py]: Unnecessary line

File: GeoCASe_functions.py [hunk 3]
```diff
@@ -75,3 +78,79 @@ def gather_data() -> dict:
         i += 1
 
     return geocase_data
+
+
+def gather_publishers() -> dict:
```
Comment (samleeflang) [inline GeoCASe_functions.py:83]: Long function, try to split it up in multiple functions

File: csv_functions.py [hunk 4]
```diff
@@ -375,8 +364,41 @@ def write_geocase_data_to_csv(geocase_data: dict):
...
+
+        writer.writerow(headers)
+
+        for v in values.values():
+            writer.writerow(v)
+
+    return csv_file
+
+
+def write_geocase_publishers_to_csv(geocase_publishers: dict):
```
Comment (samleeflang) [inline csv_functions.py]: Try to add the return type to the function

File: main.py [hunk 5]
```diff
@@ -78,7 +82,12 @@ def execute_all():
     print(f'Completed: {progress} out of {total_count}')
 
     # GeoCASe data
-    geocase_data()
+    geocase_specimens()
+    progress += 1
+    print(f'Completed: {progress} out of {total_count}')
```
Comment (samleeflang) [inline main.py]: See if you can replace the print statements to logging functions. https://docs.python.org/3/library/logging.html You can check the enrichment service for this as well

File: query_csv.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+
+        for row in reader:
+            issue_flag = row[1].lower().replace('_', ' ').capitalize()
+
+            issues_and_flags_list.append(issue_flag)
+
+    return issues_and_flags_list
+
+
+# countries = ['NL', 'DE', 'EE']
```
Comment (samleeflang) [inline query_csv.py]: Remove the out commented lines

**Review by samleeflang — COMMENTED** [review]
Tom it looks real good and a lot of work has been done.
Just a couple of general remarks:
- Try to keep functions short. Short functions are easy to read and understand. You can give them descriptive names so it tells the user what is going on
- Try to use the logging library 
- Make sure that all functions have a return type (def XXX (item:dict) -> list ) https://peps.python.org/pep-0484/

File: query_csv.py (hunk 1 above)
Comment (TomDijkema) [reply query_csv.py]: Done

File: GBIF_functions.py (hunk 2 above)
Comment (TomDijkema) [reply GBIF_functions.py]: Removed

File: GeoCASe_functions.py (hunk 3 above)
Comment (TomDijkema) [reply GeoCASe_functions.py:83]: Done, I believe

File: csv_functions.py (hunk 4 above)
Comment (TomDijkema) [reply csv_functions.py]: Done

File: main.py (hunk 5 above)
Comment (TomDijkema) [reply main.py]: Implemented in main.py, should probably implement in other places too to indicate progress

File: query_csv.py (hunk 6 above)
Comment (TomDijkema) [reply query_csv.py]: Removed

File: GeoCASe_functions.py (hunk 3 above)
Comment (samleeflang) [reply GeoCASe_functions.py:83]: The functions can be even smaller, think 10-15 lines is the size to aim for. But this is already much better.

**Review by samleeflang — APPROVED** [review]
Thanks for picking up the items. Some functions are still long and complex (nested if-else). Try to make them shorten and split up if you are creating a lot of nesting.

**Pull request merged**